### PR TITLE
Documentation

### PR DESCRIPTION
**Pull request by AnishShr**

### CODE REVIEW

File: README.asciidoc [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+- Elecronics(Design) + Programming: Anish Shrestha   
+- Firmware: Artem Kabakov   
+   
+
+== Programming
+
+=== Used Modules and Libraries
+==== Libraries
+`Numpy` for mathematical arrays and storing data for image processing operations and iterating over the objects     +
+`OpenCV` to interfce the in-built libraries for computer vision     +
```
Comment (ReikoR) [inline README.asciidoc:14]: interfce -> interface

File: README.asciidoc [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+`Pyrealsense2` to interfc the Realsense Camera and its configurations   +
```
Comment (ReikoR) [inline README.asciidoc:15]: interfc -> interface

File: README.asciidoc [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+`Pyserial` for serial communication with the STM32 microcontoller over USB      +
```
Comment (ReikoR) [inline README.asciidoc:16]: microcontoller -> microcontroller

File: README.asciidoc [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+
+==== Modules
+===== Color.py
+This program stores the data of the colors that are detected after configuring using the `config_colors.py` program.    +
+
+===== config_color.py
+This program is used mainly to configure colors for various objects in the camera frame, such as balls, court, baskets, etc.
+
+===== image_processor.py
+This is the module responsible for the image processing operations and detecting colors for image segmentation. This module produces the processed iamge frame and all of its objects necessary for further analysis and processing.
```
Comment (ReikoR) [inline README.asciidoc:28]: iamge -> image

File: README.asciidoc [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+
+===== robot_State.py
+An enum class defined to set the state at which the robot may be at an isntant.
```
Comment (ReikoR) [inline README.asciidoc:31]: isntant -> instant

File: README.asciidoc [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+The python module responsible to communicate serially with the STM32 microcontroller over USB. This module finds the USB device dynamically.
+
+===== state_machine.py
+The module which detects the state of teh robot at any current instant, taking the processed frame inputs from the `main.py` module, and acting accordingly waht to do when a robot is in a particular state, i.e., whether to advance forward, por turn, or stay still.
+
+===== main.py
+The main module which initiates the image processing module and generates a processed frame and forwards this processed frame to the `state_machine.py` module. This module is also responsible to print messages in the prompt and broadcast the current image frame (as a live stream in the screen indicating the objects, such as: largest ball, specific baskets, etc).
+
+=== Game Logic
+Initially, the robot is in the state: `FIND_A_BALL`. In this state, the robot searches the ball. If there is no ball to be seen in the processed camera frame, the it keeps rotating at the same place and detects the ball.    +
```
Comment (ReikoR) [inline README.asciidoc:43]: the it -> then it

**Review by ReikoR — CHANGES_REQUESTED** [review]

File: README.asciidoc (hunk 1 above)
Comment (AnishShr) [reply README.asciidoc]: corrected

File: README.asciidoc (hunk 2 above)
Comment (AnishShr) [reply README.asciidoc]: corrected

File: README.asciidoc (hunk 3 above)
Comment (AnishShr) [reply README.asciidoc]: corrected

File: README.asciidoc (hunk 4 above)
Comment (AnishShr) [reply README.asciidoc]: corrected

File: README.asciidoc (hunk 5 above)
Comment (AnishShr) [reply README.asciidoc]: corrected

File: README.asciidoc [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+This is the module responsible for the image processing operations and detecting colors for image segmentation. This module produces the processed iamge frame and all of its objects necessary for further analysis and processing.
+
+===== robot_State.py
+An enum class defined to set the state at which the robot may be at an isntant.
+
+===== api.py
+The python module responsible to communicate serially with the STM32 microcontroller over USB. This module finds the USB device dynamically.
+
+===== state_machine.py
+The module which detects the state of teh robot at any current instant, taking the processed frame inputs from the `main.py` module, and acting accordingly waht to do when a robot is in a particular state, i.e., whether to advance forward, por turn, or stay still.
```
[the remark this answers is not in the document]
Comment (AnishShr) [reply README.asciidoc]: corrected

File: README.asciidoc (hunk 6 above)
Comment (AnishShr) [reply README.asciidoc]: corrected

File: README.asciidoc [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+This is an implementation of a state machine logic.     +
+
+image::Images/main.PNG[]    
+
+image::Images/state_machine_workflow.PNG[]   
+
+=== Self Analysis
+- The state machine logic implementation works pretty good (not all tasks were completed: till follow the ball)
+- Robot's response to tracking the ball as per ball's movement is accurate
+- Detection of farther basket and driving towards the oppsoite half when the ball is far could be better optimized
```
[the remark this answers is not in the document]
Comment (AnishShr) [reply README.asciidoc]: corrected

File: README.asciidoc [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+image::Images/PCB_back.PNG[]
+
+
+The electronics design is completed, and the PCB was fabricated by JLCPCB and recived as well.
+
+PCB received:
+
+image::Images/PCB.PNG[]
+
+However, the board had some problems. Firstly, the pads where the battery's positive terminal needed to be connected were left unconnected. Thus, this conenction needed to be provide by some other means. Secondly, the voltage protection MOSFET kept on burning everytime.
```
[the remark this answers is not in the document]
Comment (AnishShr) [reply README.asciidoc]: corrected

File: README.asciidoc [hunk 10]
```diff
@@ -0,0 +1,295 @@
...
+- Robot's response to tracking the ball as per ball's movement is accurate
+- Detection of farther basket and driving towards the oppsoite half when the ball is far could be better optimized
+
+== Electronics
+
+The mainboard comprises of 3 DC motor drivers, each of which is responsible to control an individual DC motor. The drivers are powered by 4S LiPo abtetry and the microcontroller side is powerd through USB, whose other end is connected to the on-board PC (NUC).    +
+
+The microcontroller and the PC also communicates over serial through the same USB connection.
+
+The Electronics part comprises of two sub tasks:
```
[the remark this answers is not in the document]
Comment (AnishShr) [reply README.asciidoc]: corrected

File: README.asciidoc [hunk 11]
```diff
@@ -0,0 +1,295 @@
...
+
+=== Self Analysis
+
+- The designed PCB was good, but because of some software(KiCAD) glitches, a pad remained unconnected which could not be detected before fabrication unfortunately. This thing could have been resolved.
+- Not being able to make use of the MOSFET was a major setback for the electronics part. The were no issues with the footprint. This part could have been better dealt with.
+- The worst thing about the electronics part was that the firmware part was not complete.
+
+
+== Mechanics
+The test robot we received was in quite a bad codition. A lot of critical components and holes were either cut out or modified by mechanical means, so at first, taking the measurements for a test thrower was quite a nightmare, but after getting access to the test robot's cad files, the design of test thrower went on much more smoothly.
```
[the remark this answers is not in the document]
Comment (AnishShr) [reply README.asciidoc]: corrected

File: README.asciidoc [hunk 12]
```diff
@@ -0,0 +1,295 @@
...
+
+
+
+
+- Firstly, we made our robot quite a bit smaller (240mm in diameter)
+- The motor mounts are attached to both the top and bottom plate thus making the whole assembly more sturdy:
+
+image::Images/motor_mounts.png[]
+
+- The omniwheels were 3D printed, the rollers were pressure fitted which greatly decreases the parts complexity and increases the ease of servicability. The connecor which fastens the wheel to the motors shaft is made out of aluminium on a lathe.
```
[the remark this answers is not in the document]
Comment (AnishShr) [reply README.asciidoc]: corrected

File: README.asciidoc [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+*Georgs* Assembled the thrower and attached it to the test robot. (2h)+
+=== Monday 2022-10-24
+*Anish:* Rewired the cables including the thrower motor in test robot. Checked the connections and if the motors were well connected by powering the robot and driving the motors: wheels and thrower. Continued with the PCB design part. (1.5h)+
+
+*Georgs* Started designing the omni-wheels. Goes very slow. (3.5h)+
+
+
+=== Thursday 2022-10-27
+*Anish:* Pushed the new design to github repo. Got few more issues in the design, and solved them. Repushed the commits. Few issues were remaining. Waiting for the response/approval. (2h)+
+<<<<<<< HEAD
```
[the remark this answers is not in the document]
Comment (AnishShr) [reply README.asciidoc]: cleaned

File: README.asciidoc [hunk 14]
```diff
@@ -0,0 +1,295 @@
...
+*Georgs* Printed out all of the needed omniwheel parts (3h)+
+
+=== Monday 2022-11-21
+*Anish:* Programmed for the robot to find the ball and turn where the ball is. Used the linear scaling to map the speed from the distance between the ball and the center of the camera frame. Problems detected with linear scling: not accurat and has issues as ball is pretty far from the robot. Tried further to scale the speed, but failed. (3.5h) +
+*Artem:* Soldering the new PCB. Connected few terminals with wire as the MOSFET could not tbe interfaced. (1.5h) +
+
+=== Wednesday 2022-11-23
+*Anish:* Programmed the software part for ball detectiona nd follow. Ran into the same problem of not so precise ball tracking. Detects the ball but the tracking algorithm fails. Need to change the scaling factor and might need to incorporate PID control as the movement in left and right motion is not symmetric. Tested the thrower and calibrated the thrower speed according to the distance. interpoaltion yet to be implemented. Here also lienar scaling performed, but not precise. Reconfigured all color confugurations again as theer were few problems while ball detection. (4.5h) +
+<<<<<<< HEAD
```
[the remark this answers is not in the document]
Comment (AnishShr) [reply README.asciidoc]: cleaned

File: README.asciidoc [hunk 15]
```diff
@@ -0,0 +1,295 @@
...
+
+== Tuesday 2022-12-07
+*Georgs* All parts which need to be printed have been printed (~8h printing time)
+
+== Friday 2022-12-09
+*Georgs* Robot chassis and everything milled out with the CNC (4h)+
+
+== Saturday 2022-12-10
+*Georgs* CNCd parts cleaned up, partial assembly. Need to migrate some components from the test robot to the new one (8h+)
+>>>>>>> 73095bfb929ac4462d035fdb161416a2abb43a7f
```
[the remark this answers is not in the document]
Comment (AnishShr) [reply README.asciidoc]: Merge conflicts cleaned and blogs ordered in the correct chronological order.

File: README.asciidoc [hunk 16]
```diff
@@ -0,0 +1,295 @@
...
+image::Images/thrower.png[]
+
+
+=== Self Analysis
+
+- More thought and time should've been put into the cable management routing. In our current robot there isn't much dedicated space for wires to sit or to be managed. A great addition would be a 3d printed bracket which holds the power switches along with the battery connectors on the robots side, this would make the cable management look more tidy and organised. 
+- The small form factor makes our robot more agile on the playing field compared to our test robot. At first we had our concerns about the center of gravity being too high, but it turned out fine in the end. 
+- If we had more time for the design, we would have implemented a thrower with variable angle of throwing.
+
+=== Personal comments
```
[the remark this answers is not in the document]
Comment (AnishShr) [reply README.asciidoc:143]: Personal comments added from my side.

File: README.asciidoc [hunk 17]
```diff
@@ -0,0 +1,295 @@
...
+But, we decided not to use this PCB for the robot, and proceeded with the test robot's PCB adn firmware, and building the software programs over the test robot's electronics hardware.
+
+=== Self Analysis
+
+- The designed PCB was good, but because of some software(KiCAD) glitches, a pad remained unconnected which could not be detected before fabrication unfortunately. This thing could have been resolved.
+- Not being able to make use of the MOSFET was a major setback for the electronics part. The were no issues with the footprint. This part could have been better dealt with.
+- The worst thing about the electronics part was that the firmware part was not complete.
+
+
+== Mechanics
```
[the remark this answers is not in the document]
Comment (AnishShr) [reply README.asciidoc:104]: Robot photo added

File: README.asciidoc [hunk 18]
```diff
@@ -158,7 +159,20 @@ Suggestions for the next year students:
 
 - Time management is key, don't leave everything to the last minute.
 
+==== Anish Shrestha
 
+I mainly worked on the electronics design and software programming tasks, along with assembly of the test robot. This was my first time experience with the multi-layer PCB design, and especially the SMD components. This expereince was pretty fruitful as I learned few new things while working with the electronics design, especially dsigning the high-current paths and impedance matching in the differential pair routing. I had prior expereince with KiCAD, so the entire designing process went smooth for me.
```
Comment (ReikoR) [inline README.asciidoc:164]: * expereince -> experience
* dsigning - > designing

File: README.asciidoc [hunk 19]
```diff
@@ -158,7 +159,20 @@ Suggestions for the next year students:
...
+
+Moreover, in the software programming task, it was a new experience for me as I have very little expereince in programming before. Even though I was familiar with ROS programming, here most of the things need to be set and configured manually which was a challenging task. Also, this was my first time experience with the state machine logic, and I came to know about its significance in such as complex software architecture. Overall, it was a fun experience which taught me a lot of things.
```
Comment (ReikoR) [inline README.asciidoc:166]: expereince -> experience

File: README.asciidoc [hunk 20]
```diff
@@ -158,7 +159,20 @@ Suggestions for the next year students:
...
+
+Personally, I feel prioritizing the tasks is an undermined skill which should be acknowledge by all. This makes the flow of the work more practical, as different category of tasks have distinct deadlines. So if I could change one thing in our approach, I would put this task as a reminder. Also, configuration of the colors is something that is very crucial for this project, but is sometimes neglected which results in very poor outcomes. I needed to spend hours in the later stages of the project just configuring colors. So for an improved version of the robot, color configuration should be put a little more effort and a little more time.
```
Comment (ReikoR) [inline README.asciidoc:168]: be acknowledge -> be acknowledged

File: README.asciidoc [hunk 21]
```diff
@@ -158,7 +159,20 @@ Suggestions for the next year students:
...
+
+Also, revised cross-checking the electronics design is an important task that should be carried out. We missed out a connectionn in one of the power pads, which created a mess actually. So for next time, I would give more attention to these minute details.
```
Comment (ReikoR) [inline README.asciidoc:170]: connectionn -> connection

File: README.asciidoc [hunk 22]
```diff
@@ -158,7 +159,20 @@ Suggestions for the next year students:
...
+
+Suggestions for next year students:
+- Take some time and choose the team members wisely.
+- Plan the project timeline better.
+- Skim through the taks at the start so you would know what things you need to cover until the end of the project. If you need to acquire new skills, its better to start early.
```
Comment (ReikoR) [inline README.asciidoc:175]: taks -> tasks
its -> it's

**Review by ReikoR — CHANGES_REQUESTED** [review]

File: README.asciidoc [hunk 23]
```diff
@@ -161,18 +161,18 @@ Suggestions for the next year students:
 
 ==== Anish Shrestha
 
-I mainly worked on the electronics design and software programming tasks, along with assembly of the test robot. This was my first time experience with the multi-layer PCB design, and especially the SMD components. This expereince was pretty fruitful as I learned few new things while working with the electronics design, especially dsigning the high-current paths and impedance matching in the differential pair routing. I had prior expereince with KiCAD, so the entire designing process went smooth for me.
+I mainly worked on the electronics design and software programming tasks, along with assembly of the test robot. This was my first time experience with the multi-layer PCB design, and especially the SMD components. This expereience was pretty fruitful as I learned few new things while working with the electronics design, especially designing the high-current paths and impedance matching in the differential pair routing. I had prior experience with KiCAD, so the entire designing process went smooth for me.
```
Comment (ReikoR) [inline README.asciidoc:164]: expereience -> experience

Comment (AnishShr) [reply README.asciidoc:164]: corrected

Comment (AnishShr) [reply README.asciidoc]: All changes addressed and pushed.

File: README.asciidoc (hunk 16 above)
[the remark this answers is not in the document]
Comment (AnishShr) [reply README.asciidoc:144]: Probably no comments coming from Artem's side.

**Review by ReikoR — APPROVED** [review]

**Pull request merged**